chainer/link_hooks/timer.py

```diff
@@ -0,0 +1,138 @@
+import time
+
+import numpy
+
+from chainer.backends import cuda
+from chainer import link_hook
+
+
+class TimerHook(link_hook.LinkHook):
+    """Link hook for measuring elapsed time of :meth:`forward`.
+
+    Example:
+        Code example::
+
+            from chainer.link_hooks import TimerHook
+            hook = TimerHook()
+            with hook:
+                trainer.run()
+            hook.print_report()
+
+        Output example::
+
+              LinkName  ElapsedTime  Occurrence
+                Linear     41.42sec        2100
+                   MLP     42.09sec         700
+            Classifier     42.39sec         700
+
+        where *LinkName* is the name of link that calls the hook,
+        and *ElapsedTime* is the elapsed time the link consumed,
+        and *Occurrence* is the number of calls.
+    Warning:
+        Call graph of links are hierarchical. That means reported elapsed times
+        may be overlapping with each other and the sum may exceed the total
+        time.
+    Attributes:
+        call_history: List of measurement results. It consists of pairs of
+            the name of the link that calls this hook and the elapsed time
+            the :meth:`forward` method of link consumes.
+    """
+
+    name = 'TimerHook'
+
+    def __init__(self):
+        self.call_history = []
+        self._running_stack = []
+        self._depth = 0
+        self._total_time = 0
+
+    def _preprocess(self):
+        if self.xp is numpy:
+            start = time.time()
+            self._running_stack.append(start)
+        else:
+            assert self.xp is cuda.cupy
+            start = cuda.Event()
+            stop = cuda.Event()
+            start.record()
+            self._running_stack.append((start, stop))
+        self._depth += 1
+
+    def forward_preprocess(self, args):
+        self.xp = args.link.xp
+        self._preprocess()
+
+    def _postprocess(self, link):
+        if self.xp is numpy:
+            start = self._running_stack.pop()
+            stop = time.time()
+            elapsed_time = stop - start
+        else:
+            assert self.xp is cuda.cupy
+            start, stop = self._running_stack.pop()
+            stop.record()
+            stop.synchronize()
+            # Note that `get_elapsed_time` returns result in milliseconds
+            elapsed_time = cuda.cupy.cuda.get_elapsed_time(
+                start, stop) / 1000
+        self.call_history.append((link.__class__.__name__, elapsed_time))
+
+        assert self._depth > 0
+        self._depth -= 1
+        if self._depth == 0:
+            self._total_time += elapsed_time
+
+    def forward_postprocess(self, args):
+        link = args.link
+        assert link.xp == self.xp
+        self._postprocess(link)
+
+    def total_time(self):
+        """Returns total elapsed time in seconds."""
+        return self._total_time
+
+    def summary(self):
+        """Returns a summary of time profiling in links.
+        Returns:
```

Comment:
```suggestion

         Returns:
```

Could you fix the following error?
```
Warning, treated as error:
/home/travis/build/chainer/chainer/chainer/link_hooks/timer.py:docstring of chainer.link_hooks.TimerHook.summary:3:Unexpected indentation.
```